### PR TITLE
[feature-wip](new-scan) add scanner scheduling framework

### PR DESCRIPTION
**Pull request by morningman**
# Proposed changes

Issue Number: close #xxx

## Problem summary

There are currently many types of ScanNodes in Doris. And most of the logic of these ScanNodes is the same, including:

1. Runtime filter
2. Predicate pushdown
3. Scanner generation and scheduling

So I intend to unify the common logic of all ScanNodes.
Different data sources only need to implement different Scanners for data access.
So that the future optimization for scan can be applied to the scan of all data sources,
while also reducing the code duplication.

This PR mainly adds 4 new class:

1. VScanner
     All Scanners' parent class. The subclasses can inherit this class to implement specific data access methods.

2. VScanNode
     The unified ScanNode, and is responsible for common logic including RuntimeFilter, predicate pushdown, Scanner generation and scheduling.

3. ScannerContext
    ScannerContext is responsible for recording the execution status
    of a group of Scanners corresponding to a ScanNode.
    Including how many scanners are being scheduled, and maintaining
    a producer-consumer blocks queue between scanners and scan nodes.

    ScannerContext is also the scheduling unit of ScannerScheduler.
    ScannerScheduler schedules a ScannerContext at a time,
    and submits the Scanners to the scanner thread pool for data scanning.

4. ScannerScheduler
    Unified responsible for all Scanner scheduling tasks

Test:
This work is still in progress and default is disabled.
I tested it with jmeter with 50 concurrency, but currently the scanner is just return without data.
The QPS can reach about 9000.
I can't compare it to origin implement because no data is read for now. I will test it when new olap scanner is ready. 

TODO:
1. Gradually migrate all existing ScanNodes to VScanNode.

## Checklist(Required)

1. Does it affect the original behavior: 
    - [ ] Yes
    - [x] No
    - [ ] I don't know
2. Has unit tests been added:
    - [ ] Yes
    - [x] No
    - [ ] No Need
3. Has document been added or modified:
    - [ ] Yes
    - [ ] No
    - [x] No Need
4. Does it need to update dependencies:
    - [ ] Yes
    - [x] No
6. Are there any changes that cannot be rolled back:
    - [ ] Yes (If Yes, please explain WHY)
    - [x] No

## Further comments

If this is a relatively large or complex change, kick off the discussion at [dev@doris.apache.org](mailto:dev@doris.apache.org) by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

File: be/src/vec/exec/scan/scanner_scheduler.h [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+    void _scanner_scan(ScannerContext* ctx, VScanner* scanner);
+
+private:
+    // scheduling queue number.
+    // TODO: make it configurable.
+    static const int QUEUE_NUM = 4;
+    // the ScannerContext will be submitted to the scheduling queue roundrobin.
+    // _queue_idx pointer to the current queue.
+    std::atomic_int _queue_idx = {0};
+    BlockingQueue<ScannerContext*>** _scheduling_queues;
```
Comment (Jibing-Li) [inline be/src/vec/exec/scan/scanner_scheduler.h:68]: Why do we need multiple scheduling queues?

File: be/src/vec/exec/scan/scanner_context.cpp [hunk 2]
```diff
@@ -0,0 +1,123 @@
...
+    // The calculation logic is as follows:
+    //  1. Assuming that at most M rows can be scanned in one scan(config::doris_scanner_row_num),
+    //     then figure out how many blocks are required for one scan(block_per_scanner).
+    //  2. The maximum number of concurrency * the blocks required for one scan,
+    //     that is, the number of blocks that need to be pre-allocated
+    auto doris_scanner_row_num =
+            limit == -1 ? config::doris_scanner_row_num
+                        : std::min(static_cast<int64_t>(config::doris_scanner_row_num), limit);
+    int real_block_size =
+            limit == -1 ? block_size : std::min(static_cast<int64_t>(state->batch_size()), limit);
```
Comment (Jibing-Li) [inline be/src/vec/exec/scan/scanner_context.cpp:38]: the variable `block_size` in ` limit == -1 ? block_size ` is not initialized, should we use `state->batch_size()` instead?

**Review by Jibing-Li — DISMISSED** [review]
LGTM

**Review by Jibing-Li — DISMISSED** [review]

**Review by yiguolei — APPROVED** [review]
LGTM

**Pull request merged**